File: code/python/84.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        heights.append(0)
        stack = []
        i = 0
        #维护一个单调递增栈
        while i<len(heights):
            if len(stack)==0 or heights[i]>heights[stack[-1]]:
                stack.append(i)
                i+=1
            #触发弹出条件，要计算当前最大的矩形大小
            else:
                cur = stack[-1]
                stack.pop()
                res = max(res,heights[cur]*(i if len(stack)==0 else i-stack[-1]-1))
        return res

File: code/java/84.java

class Solution {
    public int largestRectangleArea(int[] heights) {
        int[] heightsWithZero = new int[heights.length+1];
        heightsWithZero[heights.length] = 0;
        for(int i=0;i<heights.length;i++) heightsWithZero[i] = heights[i];

        Stack stack = new Stack();
        int res = 0;
        int index = 0;

        ////System.out.println(Arrays.toString(heightsWithZero));

        while(index<heightsWithZero.length){
            if(stack.empty()||heightsWithZero[index]>heightsWithZero[(int)stack.peek()]){
                stack.push(index);
                index+=1;
            }
            else{
                int cur = (int)stack.pop();
                res = Math.max(res,heightsWithZero[cur]*(stack.size()==0?index:index-(int)stack.peek()-1));
            }
        }
        return res;
    }
}